[mission001] todolist crud

ganeodolu opened the pull request:
#1 Mission001

- [x]  todo list에 todoItem을 키보드로 입력하여 추가
- [x]  todo list의 체크박스를 클릭하여 complete 상태로 변경. (li tag 에 completed class 추가)
- [x]  todo list의 x버튼을 이용해서 해당 엘리먼트를 삭제
- [x]  todo list를 더블클릭했을 때 input 모드로 변경. (li tag 에 editing class 추가) 단 이때 수정을 완료하지 않은 상태에서 esc키를 누르면 수정되지 않은 채로 다시 view 모드로 복귀
- [x]  todo list의 item갯수를 count한 갯수를 리스트의 하단에 보여주기
- [x]  todo list의 상태값을 확인하여, 해야할 일과, 완료한 일을 클릭하면 해당 상태의 아이템만 보여주기

필터버튼 테두리 유지를 구현하면서 selected가 적용되는 부분은 동준님 코드를 참조하였습니다. 필터기능 구현시 app에서 setState대신에 render메소드를 사용하여 this.data가 직접 수정되는 것을 방지하였습니다.

미션2 사용했던 제 코드와 로토님 라이브코딩시 사용했던 코드를 기반으로 작성하였습니다. 스터디했었던 난이도와 비슷하여 막막하지 않고, 미션 해결될 때마다 즐거웠습니다.   
기능구현하기에도 녹록치 않아서 컴포넌트별로 기능이 잘 구분되지는 않은 것 같습니다.
다른 분들 코드 참조하면서 발전시키도록 하겠습니다.
한번 봐주시고 피드백 부탁드립니다. :pensive:

ganeodolu committed:
todolist에 todoitem을 엔터키로 추가

ganeodolu committed:
M1-1 todolist 입력 수정

CSS형식에 맞게 todolist의 타겟 변경,
innerHTML 태그 추가

ganeodolu committed:
M1 2~5 기능구현

2. 토글버튼으로 할일 completed상태변경
3. 삭제버튼으로 할일 삭제
4. 할일 더블클릭시 에디트모드
5. 할일 개수 하단에 표시

ganeodolu committed:
M1 삭제기능 보완

data index 위치를 toggle 태그안에서 li 태그로 변경

eastjun-dev commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,62 @@
+
+    const $todoList = document.querySelector('.todo-list')
+    const todoList = new TodoList($todoList, this.data,
+        (index) => {
```
parameter로 함수들을 넘기는데, 함수들이 이름이 없어서 무슨 역할을 하는 함수인지 모든 코드를 읽어봐야 알 수 있는 것 같아요.
```javascript
{
functionName: () => {},
functionName: () => {},
}
```
위와 같은 형태로 parameter를 넘기는게 가독성 측면에서 더 좋을 것 같습니다 :)

eastjun-dev commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,62 @@
+    const $todoInput = document.querySelector('.new-todo')
+    const $todoEdit = document.querySelector('.todo-list')
+    const todoInput = new TodoInput($todoInput, $todoEdit,
+        (text) => {
```
14 line과 같은 맥락으로 function name이 있는 key: value 형태이면 더 좋을 것 같아요 :)

ganeodolu replied:
객체가 익숙치 않아서 사용 안했었는데 사용하니 훨씬 보기가 편하네요.

eastjun-dev commented on mission001/ganeodolu/js/TodoCount.js:
```diff
@@ -0,0 +1,31 @@
+    this.$targetCount = $targetCount;
+    this.$targetCountFilter = $targetCountFilter;
+
+    // 새로운 데이터를 불러서 다시 렌더링하는 setState함수
```
꼭 필요한게 아니면 주석은 없어도 될 것 같아요 :)

ganeodolu replied:
신경 쓴다고 주석을 달았는데 굳이 필요한게 아니었군요. 모두 삭제하였습니다.

eastjun-dev commented on mission001/ganeodolu/js/TodoInput.js:
```diff
@@ -0,0 +1,9 @@
+function TodoInput($targetInput, $targetEdit, onAdd) {
+    // 입력창에서 엔터키 입력시 할일추가
```
꼭 필요한게 아니면 주석은 없어도 될 것 같아요 :)

eastjun-dev commented on mission001/ganeodolu/js/TodoInput.js:
```diff
@@ -0,0 +1,9 @@
+function TodoInput($targetInput, $targetEdit, onAdd) {
+    // 입력창에서 엔터키 입력시 할일추가
+    $targetInput.addEventListener('keydown', (e) => {
+        if (e.keyCode === ENTER_KEY_CODE) {
```
`event.keyCode`같은 경우 [MDN 문서](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode)를 보시면 deprecated된 것을 확인하실 수 있어요~! [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode)를 참고해서 업데이트 하시면 좋을 것 같아요 :)

ganeodolu replied:
따로 비교할 상수를 만들지 않아도 바로 버튼명과 비교가능하군요. 잘 사용하겠습니다.
`e.code === "Enter"`

eastjun-dev commented on mission001/ganeodolu/js/TodoInput.js:
```diff
@@ -0,0 +1,9 @@
+    $targetInput.addEventListener('keydown', (e) => {
+        if (e.keyCode === ENTER_KEY_CODE) {
+            onAdd($targetInput.value)
+            $targetInput.value = ''; // 입력 후 입력란이 공란이 되도록 함
```
꼭 필요한게 아니면 주석은 없어도 될 것 같아요 :)

eastjun-dev commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+    if (this === window) {
+        throw new Error(error.NO_USED_NEW_KEYWORD)
+    }
+    else if (Array.isArray(this.data) == false) {
```
자바스크립트에서 값을 비교할 때 특수한 경우가 아니면 `===` 로 비교해야 예상치 못한 오류가 발생할 확률이 줄어들 것 같습니다 :)

ganeodolu replied:
주의하도록 하겠습니다. :grin:

eastjun-dev commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+    else if (Array.isArray(this.data) == false) {
+        throw new Error(error.NOARRAY_DATA)
+    }
+    // 토글버튼과 삭제버튼 기능구현
```
꼭 필요한게 아니면 주석은 없어도 될 것 같아요 :)

eastjun-dev commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+            onRemoveClick(index)
+        }
+    })
+    // 할일 더블클릭시 에디트 모드로 변경
```
꼭 필요한게 아니면 주석은 없어도 될 것 같아요 :)

eastjun-dev commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+    // 할일 더블클릭시 에디트 모드로 변경
+    this.$target.addEventListener('dblclick', (e) => {
+        const { className } = e.target;
+        const { index } = e.target.parentNode.parentNode.dataset
```
ui에 나중에 tag가 추가되는 경우가 발생할 수도 있을 것 같은데 parentNode보다는 [closest](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)를 이용해보면 어떨까요~?

ganeodolu replied:
closest 는 처음 보았는데 유용하게 쓰일 것 같네요.

eastjun-dev commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+        // renderHTMLText 변수 안에 data객체안의 text의 value 값에 태그를 붙여서 저장
+        const renderHTMLText = this.data.map((val, idx) => {
+            // const TOGGLE_INPUT = `<input class="toggle" type="checkbox" data-index=${idx} ${val.isCompleted ? 'checked' : ""}>`;
+            const TOGGLE_INPUT = `<input class="toggle" type="checkbox"  ${val.isCompleted ? 'checked' : ""}>`;
```
html string을 tag 나누신 이유가 있을까요?
각 tag를 나타내는 변수를 외부에서 재사용하는 경우가 없기 때문에, 아래처럼 하나의 template변수로 관리하는게 html tag 구조를 파악하기 더 좋을 것 같아요 :)

```javascript
const todoItemTemplate = () => `
     <li>
        <div class="view">
          <input class="toggle" type="checkbox">
          <label class="label">${todoItem.contents}</label>
          <button class="destroy"></button>
        </div>
        <input class="edit" value="${todoItem.contents}">
     </li>`
```

ganeodolu replied:
태그를 일렬로 표현하다보니 짧게 표현해야겠다는 생각으로 태그단위로 잘랐었습니다. innerHTML로 표현시 이 방법 훨씬 가독성이 좋네요. 감사합니다.

eastjun-dev commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+            // // isCompleted 가 참이면 <strike>태그를 넣어서 저장(완료되었다는 의미), false면 <div>만 넣어서 저장
+            return `<li ${val.isCompleted ? 'class="completed"' : (val.isCompleted === false && val.isEditing === false) ? "" : 'class="editing"'} data-index=${idx}><div class="view">${TOGGLE_INPUT}<label class="label">${val.text}</label>${DESTROY_BUTTON}</div>${EDIT_INPUT}</li>`;
+        }).join('');
+        this.$target.innerHTML = `${renderHTMLText}`
```
`renderHTMLText` 변수에 있는 값을 this.$target.innerHTML에 그대로 넣는 것이면
`this.$target.innerHTML = renderHTMLText`과 같이 표현해도 괜찮지 않을까요?

ganeodolu replied:
처음에 태그를 같이 쓰다가 변수만 남았는데도 불필요하게 `${}`는 남겼었네요. 삭제하였습니다.

amorfati0310 replied:
변수명이 renderHTMLText 여서 함수 같은데  결과를 저장한 변수라면 renderedHTMLText 로 하시면 더 좋을 것 같네요

ganeodolu replied:
조언대로 변수명을 변경하였습니다. 여러 코드 보면서 변수명 명명시 참고하겠습니다.

eastjun-dev commented on mission001/ganeodolu/js/TodoList.txt:
```diff
@@ -0,0 +1,56 @@
+function TodoList($target, data, onToggleClick, onTodoDoubleClick, onRemoveClick){
```
저도 종종 하는 실수인데, 어플리케이션과 관계 없는 파일은 원격 Repository에 올리지 않는게 좋은 것 같아요 :)

ganeodolu replied:
질문드릴려고 저장했다가 같이 업로드 되버렸네요. 주의하겠습니다.

eastjun-dev commented on mission001/origin/.vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
보통 IDE의 설정파일은 원격 저장소에 관리하지 않아요~ [.gitIgnore](https://www.google.com/search?client=firefox-b-d&q=gitIgnore)를 이용해서 실수로 push하지 않도록 세팅해놓으시면 훨씬 편하실 거에요 :)

ganeodolu replied:
생각도 안해봤었던 문제였어요. 잘 설정해두겠습니다.

amorfati0310 replied:
.env 랑 dot 도 같이 알아보시면 좋을 것 같습니다 👍

ganeodolu replied:
.env는 환경설정 파일이라 이해가 가지만, dot파일은 찾아보니 그래프 관련 파일 또는 단축키 관련 파일이라고 나오는데 사용해본적이 없어서 잘 모르겠습니다. 어떤 dot파일인지 알려주실 수 있나요??

eastjun-dev reviewed (changes requested):
프로그래머스에서 배우신 부분들을 꼼꼼하게 다시 복습하시면서 구현해주셔서 저도 다시 복습하는 시간이 되었습니다 
😄  너무너무 고생하셨어요~!
지금도 좋은데 제 기준에서 드릴 수 있는 피드백을 코멘트로 남겼는데요. 크게 2가지를 조금 더 신경써주시면 훨씬 좋을 것 같아요~!

## 1. 필요없는 주석 제거

주석은 꼭 필요한 경우가 아니면 제거해주시는게 좋아요~! 주석이 필요하다는건 코드도 읽고, 주석도 읽어야 하기 때문에 더욱 복잡하게 되는 것 같아요. 특히 협업하는 경우엔 더더욱 주석은 필요한 경우만 남겨야 서로 원활한 커뮤니케이션을 하는데 도움이 될 것 같아요 :)

## 2. 역할에 따른 함수 분리

1개의 함수는 1가지의 기능만 하도록 최대한 분리해야 나중에 재사용성이 훨씬 높아지고, 보다 구조적으로 설계할 수 있을 것 같아요. TodoList.js의 render함수를 보시면 templateHTML도 만들어주고, validation 체크도 하고 2가지 기능을 한번에 하는데 이런 부분들을 역할을 고민해보고 분리해서 불러오는 방식으로 만드는게 더 좋을 것 같아요 :)

YongHoonJJo commented on mission001/ganeodolu/js/index.js:
```diff
@@ -0,0 +1,23 @@
+    INVALID_DATA: "data타입이 문자열이 아닙니다.",
+};
+const ENTER_KEY_CODE = 13;
+const ESC_KEY_CODE = 27;
```
키보드의 키 값들도 위의 error 처럼 한번에 묶어서 관리하는 것은 어떨까요?

ganeodolu replied:
객체가 익숙하지 않아서 피해서 사용해왔는데 익숙해지도록 하겠습니다.
e.KeyCode 대신에 e.code로 변경하여 키값들은 삭제하였습니다.

YongHoonJJo replied:
`e.code` 보다는 `e.key` 가 더 명확하지 않을까 싶어요..!! 엔터랑  ESC 는 같은 값이라 상관 없는데, 알파벳 같은 경우는 차이가 납니다.

<https://keycode.info> 

위에서 확인해보세요!! :)

ganeodolu replied:
할일에 한글입력시 중복 입력되는 이슈가 있어서 e.code에서 e.key로 변경하였습니다. 그리고 조언대로 역시 묶어서 관리하는게 좋을 것 같아서 수정하였습니다.

YongHoonJJo commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,62 @@
+
+    const $todoList = document.querySelector('.todo-list')
+    const todoList = new TodoList($todoList, this.data,
+        (index) => {
```
저도 동준님과 같은 생각입니다..!

ganeodolu replied:
네 변경하니 눈에 더 잘 들어오네요. 감사합니다.

YongHoonJJo commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,62 @@
+        todoList.setState(this.data)
+        todoCount.setState({
+            totalCount: this.data.length,
+            completedCount: this.data.filter(todo => todo.isComplete).length
```
`todo.isComplete` 오타인가요..? ㅎㅎ

ganeodolu replied:
기존 코드를 변경없이 가져왔었네요. 수정하도록 하겠습니다. ^^

YongHoonJJo commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+        }
+    })
+
+    this.render = function () {
```
객체가 공통적으로 사용하는 메서드의 경우, prototype 을 통해서 저장하는 것은 어떨까요?

ganeodolu replied:
prototype에 대해 이해가 부족하여 너무 늦게 답변을 드렸네요. 말씀하신데로 TodoList.prototype.render를 새로 만들어 수정하였습니다. 그런데 TodoList함수에서 분리하여 prototype을 따로 사용할 때 메모리 효율외에 어떤 장점이 있는지 알려주실 수 있을까요??

YongHoonJJo commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+            const TOGGLE_INPUT = `<input class="toggle" type="checkbox"  ${val.isCompleted ? 'checked' : ""}>`;
+            const DESTROY_BUTTON = '<button class="destroy"></button>';
+            const EDIT_INPUT = `<input class="edit" value="${val.text}">`
+            if (!val.text) {            // data.text 값이 있는지 확인
```
예외처리를 하기 전에, 이미 위의 라인에서 `val.text` 의 값을 사용하고 있습니다. 값을 사용하기 이전에 예외처리를 하는게 맞지 않을까요?

ganeodolu replied:
임시방편으로 변수를 만들다 보니 저렇게 되었네요. 예외처리 이후로 변경하였습니다.  ^^;;

YongHoonJJo commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,62 @@
+        (index) => {
+            const nextData = [...this.data]
+            if (!nextData[index].isCompleted) {
+                nextData[index].isEditing = nextData[index].isEditing ? false : true;
```
nextData[index].isEditing = !nextData[index].isEditing
이렇게 하셔도 될 것 같습니다!! :)

ganeodolu replied:
훨씬 간단하게 표현되네요. 수정하였습니다!

YongHoonJJo commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+        this.render()
+    }
+
+    if (this === window) {
```
new 키워드로 생성자 함수를 만들었는지 체크한 부분을 좋은 것 같습니다. 하지만, 이 예외처리가 다른 생성자 함수들 내부에는 적용되어 있지 않은데, 여기에서만 예외처리를 하신 이유가 따로 있으신가요?

ganeodolu replied:
생각없이 기존 코드를 재사용하였는데, 꼭 집어서 말씀해주시니 코드의 의미를 다시 보게 되었습니다. 다른 생성자 함수에도 적용하였습니다.

YongHoonJJo commented on mission001/ganeodolu/js/TodoInput.js:
```diff
@@ -0,0 +1,9 @@
+function TodoInput($targetInput, $targetEdit, onAdd) {
```
`$targetEdit` 이 사용된 곳이 보이지 않습니다. 그리고 new 키워드를 통해 App.js 에서 호출하였기 때문에 컴포넌트로써 만드신 것 같아 보입니다. 하지만 TodoInput() 의 내용은 단순히 이벤트를 등록하는 역할만 하는 것으로, 생성자 함수라기 보단, 일반 함수처럼 보입니다. 어떤 대안을 제시하기도 조금은 애매합니다만, 어떻게 생각하시나요..? ㅎㅎ

ganeodolu replied:
첫번째, $targetEdit는 코멘트처럼 함수에서 사용되지 않아서 삭제하였습니다. 두번째, TodoList파일을 기반으로 TodoInput을 만들었으나 말씀해주신데로 굳이 생성자를 사용하지 않아도 될 것 같아서 App.js에서 new를 삭제하였습니다. 다시 생각하게 되는 코멘트 감사합니다.

YongHoonJJo commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,62 @@
+    })
+
+    const $todoInput = document.querySelector('.new-todo')
+    const $todoEdit = document.querySelector('.todo-list')
```
위에서 이미 `const $todoList = document.querySelector('.todo-list')` 와 같이 선언한적이 있습니다. 추가로 하셔야 할 이유가 있을까요?

ganeodolu replied:
각 함수 단위로만 생각하다가 중복이 되는지 몰랐었네요. 삭제하도록 하겠습니다.

YongHoonJJo reviewed (changes requested):
고생하셨습니다! 저도 많이 부족한 실력이지만, 제가 가지고 있는 지식 내에서 이렇게 해보는 것을은 어떨까 싶은 것들에 대해 코멘트 남겨봤습니다..!!

ganeodolu committed:
M1 리뷰 일부반영

객체화, e.code 변경, 주석삭제

ganeodolu committed:
M1 리뷰 반영

html string 수정, 예외처리

프로토타입제외

YongHoonJJo commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+        const { className } = e.target;
+        const { index } = e.target.parentNode.dataset;
+        if (className === 'edit') {
+            if (e.code === "Enter") {
```
e.code 보다는 e.key 가 더 명확하지 않을까 싶어요..!! 엔터랑 ESC 는 같은 값이라 상관 없는데, 알파벳 같은 경우는 차이가 납니다.

https://keycode.info

위에서 확인해보세요!! :)

ganeodolu replied:
KeyboardEvent.keyCode MDN문서에서 event.code를 추천해서 두가지만 있는 줄 알았는데 event.key도 존재하는군요. 알파벳 같은 경우에 용훈님 설명처럼 간단하게 표현되네요.
key : s, code : keyS
key로 변경하였습니다.
keyCode, key, code 이름이 비슷해서 헷갈리네요 😨

ganeodolu committed:
M1-6필터기능 구현중

amorfati0310 commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,103 @@
+function App(data) {
+    this.data = data;
+    // this.render = function (filteredData){
```
요런 주석은 pr 보내기 전에 제거 되면 더 좋겠네요 :)

ganeodolu replied:
필터를 아직 구현중이라 주석이 남아있었습니다. 
구현을 완료하면 삭제하도록 하겠습니다.

amorfati0310 commented on mission001/ganeodolu/js/App.js:
```diff
@@ -0,0 +1,103 @@
+        $target: $todoList,
+        $targetFilter: $todoFilter,
+        data: this.data,
+        onToggleClick: (index) => {
```
이벤트 핸들러들 이름이 일관성 있어서 좋네요 👍  on prefix 
다만  `onEditTodo , onChangeTodo on 동사 + 명사` on 붙여진 부분도 더 일관성있게 다루면 어떨까 싶습니다

ganeodolu replied:
로토님이 이벤트와 연관된 함수에는 보통 on을 앞에 붙힌다고 사용하고 있었습니다.
명명규칙이 통일되지 않았었네요. on 동사 + 명사 형태로 수정하였습니다.

amorfati0310 commented on mission001/ganeodolu/js/TodoCount.js:
```diff
@@ -0,0 +1,34 @@
+        this.render();
+    }
+
+    if (this === window) {
```
(this === window) instanceOf 를 활용하시면 더 가독성 좋은 코드일 것 같아요 
this는 실행 컨텍스트에 따라 달라져서  ex)
```js
// 조금 억지긴 하지만 
const app.todoCount = TodoCount
app.todoCount({...todoCountdata})
객체 프로퍼티로 실행되면 this가 app을 가리켜서
 일반 function으로 그냥 호출했을때만 this가 전역 객체를 바라봅니다.
bind, call, apply 로도 this가 바뀔 수가 있어서 
new keyword가 아닌경우 instance로 바로 검출을 하시는게 더 좋을 것 같습니다.
```

ganeodolu replied:
`if(!todoList instanceof TodoList || !todoCount instanceof TodoCount){
        throw new Error(error.NO_USED_NEW_KEYWORD)
    }`
App 마지막 줄에 instance를 판별하는 구문을 넣어보았습니다. 하지만 new를 사용하지 않으면 인스턴스를 확인하는 줄까지 오기전에 $target이 undefined라는 에러가 나와서 입력한 에러메시지는 나오지 않습니다. 적절한 위치가 있을까요?

amorfati0310 commented on mission001/ganeodolu/js/TodoCount.js:
```diff
@@ -0,0 +1,34 @@
+    }
+    this.render();
+
+}
```
아 궁금해서 이거는 그냥 남겨요  
저 빨간 엔터 No new line end of line 저 부분 없애는 방법이 있는 것 같은데  ㅎㅎ 알게되면 공유드리겠습니다 
공백이 있어서 그런걸까요?

ganeodolu replied:
빨간 엔터가 어떤 부분을 말하는 건지 잘 모르겠어서요. 혹시 다시 설명해주실 수 있나요??

ganeodolu replied:
이제야 빨간엔터 의미를 알았네요. 호석님 리뷰를 보니 마지막 줄에 개행을 안해서 그런건가봐요.

amorfati0310 commented on mission001/ganeodolu/js/index.js:
```diff
@@ -0,0 +1,23 @@
+const error = {
```
실행부분과 상수 부분을 따로 관리하면 좋을 것 같네요 👍

ganeodolu replied:
constant.js로 상수를 분리하였습니다.

amorfati0310 commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,83 @@
+        const { className } = e.target;
+        // const { index } = dataset
+        // console.log(e.target)
+        if (className === 'all selected') {
```
if else 가 많아지면  switch case 혹은 객체로 빼서 대응하는 함수에 맵핑 시키면 가독성이 더 좋아지는데 시도해보시면 좋을 것 같아요 :)

ganeodolu replied:
이 부분에서는 다른 로직으로 변경해서 사용하지는 않았지만 다른 부분에서 사용해 보았습니다. 
`if (className === 'toggle') {
            onToggleClick(index)
        } else if (className === 'destroy') {
            onRemoveClick(index)
        }`
위 코드를 아래코드처럼 수정
`        switch (className) {
            case 'toggle': onToggleClick(index) 
                break;
            case 'destroy': onRemoveClick(index)
                break;
        }`
다음부터 switch문도 잘 참고하도록 하겠습니다.

amorfati0310 commented on mission001/ganeodolu/index.html:
```diff
@@ -0,0 +1,65 @@
+    </ul>
+  </div>
+</section>
+<script src="js/App.js"></script>
```
이렇게 파일 하나 하나 script를 따로 import 해야 되는 부분을 어떻게 해결할 수 있을지 고민해보시는 것도 좋을 것 같네요 :)

ganeodolu replied:
import, export를 활용하여 module로 사용할 index.js만 사용하도록 변경하였습니다.

amorfati0310 reviewed (commented):
수고 많으셨습니다 리뷰가 도움이 되시기를 👍

amorfati0310 commented on mission001/ganeodolu/js/TodoList.js:
```diff
@@ -0,0 +1,68 @@
+                throw new Error(error.NOT_DATA)
+            }
+            // data.text 내용이 문자열이 아닌 경우를 데이터가 이상하다고 가정
+            else if (typeof (val.text) != "string") {
```
비교연산자는 === or !==  으로 해주시면 좋을 것 같아요 :)

ganeodolu replied:
네 변경하였습니다. ===은 익숙한데 !==는 자꾸 != 이렇게 쓰게 되네요.

amorfati0310 reviewed (commented):
다른 분들 리뷰보다가 추가로 보이는 게 있어서 추가 코멘트를 남겨요 도움이 되시기를 바랍니다 ~~

eastjun-dev reviewed (approved):
피드백 요청드린 부분 세심하게 확인해주시고 반영해주셔서 피드백 과정이 즐거웠습니다 :)
호준님만의 속도로 계속해서 정진해나가면서 성장하시길 바랍니다 화이팅! 👏👏👏
저는 approve 진행할게요 :)

ganeodolu committed:
M1-6 필터기능 구현

Happy 2020 Year !!!

ganeodolu committed:
M1 리뷰 반영

1. if else를 switch로 변환
2. import, export 사용
3. instanceOf 사용하여 instance 검사

프로토타입 미반영

ganeodolu committed:
M1 prototype 사용 및 개행추가
